zkvm/docs/zkvm-blockchain.md

```diff
@@ -0,0 +1,406 @@
+The output is a single 32-byte hash value.
+The input list must be prefix-free;
+that is,
+no element can be a prefix of any other.
+Given a sorted list of n unique inputs,
+
+```
+D[n] = {d(0), d(1), ..., d(n-1)}
+```
+
+the MPTH is thus defined as follows.
+
+The hash of an empty Merkle patricia tree list is a 32-byte challenge string with the label `patricia.empty`:
+
+```
+MPTH(T, {}) = T.challenge_bytes("patricia.empty")
+```
+
+To compute the hash of a list with one entry,
+commit it to the transcript with the label `"patricia.leaf"` and then generate a 32-byte challenge string with the same label:
+
+```
+T.commit("patricia.leaf", d(0))
+MPTH(T, {d{0)}) = T.challenge_bytes("patricia.leaf")
+```
+
+To compute the hash of a list with two or more entries:
+1. Let the bit string `p` be the longest common prefix of all entries;
+2. Let k be the number of items with prefix `p||0`
+   (that is,
+   `p` concatenated with the single bit 0).
+3. Let L be recursively defined as `MPTH(T, D[0:k])`
+   (the hash of the first `k` elements of D).
+4. Commit `L` to `T` with the label `"patricia.left"`.
+5. Let R be recursively defined as `MPTH(T, D[k:n])`
+   (the hash of the remaining `n-k` elements of D).
+5. Commit `R` to `T` with the label `"patricia.right"`.
+6. Generate a 32-byte challenge string with the label `"patricia.node"`.
+
+```
+T.commit("patricia.left", MPTH(T, D[0:k]))
+T.commit("patricia.right", MPTH(T, D[k:n]))
+MPTH(T, D) = T.challenge_bytes("patricia.node")
+```
+
+
+# Procedures
+
+In the descriptions that follow,
+the word “verify” means to test whether a condition is true.
+If it’s false,
+all pending procedures abort and a failure result is returned.
+
+## Start new network
+
+A node starts here when creating a new blockchain network.
+Its [blockchain state](#blockchain-state) is set to the result of the procedure.
+
+Inputs:
+- `timestamp_ms`,
+  the current time as a number of milliseconds since the Unix epoch:
+  00:00:00 UTC Jan 1,
+  1970.
+- `refscount`,
+  the number of recent block ids to cache for
+  [nonce](zkvm-spec.md#nonce)
+  uniqueness.
+
+Output:
+- Blockchain state.
+
+Procedure:
+1. [Make an initial block](#make-initial-block)
+   `initialblock` from `timestamp_ms` and `refscount`.
+2. Return a blockchain state with its fields set as follows:
+   - `initialheader`:
+     `initialblock.header`
+   - `tipheader`:
+     `initialblock.header`
+   - `utxos`:
+     empty set
+   - `nonces`:
+     empty set
+   - `refids`:
+     empty set
+
+## Make initial block
+
+Inputs:
+- `timestamp_ms`,
+  the current time as a number of milliseconds since the Unix epoch:
+  00:00:00 UTC Jan 1,
+  1970.
+- `refscount`,
+  the number of recent block ids to cache for
+  [nonce](zkvm-spec.md#nonce)
+  uniqueness.
+
+Output:
+- a
+  [block](#block).
+
+Procedure:
+1. Compute `txroot`,
+   the root hash of an empty
+   [Merkle binary tree](zkvm-spec.md#merkle-binary-tree)
+   with label `transactions`.
```

Comment:
Those labels are repeated across the spec, maybe we need separate sections like we do with [Transaction ID](https://github.com/interstellar/slingshot/blob/main/zkvm/docs/zkvm-spec.md#transaction-id) in ZkVM? 

```
## Transaction root

Transaction ID is defined as a [merkle hash](#merkle-binary-tree) of a list consisting of 
a [header entry](#header-entry) followed by all the entries from the [transaction log](#transaction-log):

    T = Transcript("ZkVM.txroot")
    txroot = MerkleHash(T, {tx[0], tx[1], ...})
```

Each transaction is committed to the [transcript](#transcript) via its [transaction ID](zkvm-spec.md#transaction-id):

    T.commit("tx", tx.id)

```

Note that for trees we need to specify how to commit the items.

Similarly, for `## UTXO root` and `## Nonce root`. And then we can xref to these sections throughout the doc.

Reply:
Good call.